### PR TITLE
+act #22109 add documentation annotations in akka.annotation

### LINKED ISSUE #22109: Introduce API stability annotations
Originated in https://github.com/akka/akka-http/issues/438 

Let's start simple:

- `@InternalApi` - equivalent of current `/** INTERNAL API*/` comments, we'll work on MiMa to be aware of these (!) so no filtering out when touching such classes/methods will be needed
- `@ApiMayChange` - To mark new "experimental" APIs, such as the #662 API

These annotations will land immediately in akka.actor, we'll use them from there in Akka HTTP.

Hope this sounds good, cheers!

### PR DESCRIPTION
**Pull request by ktoso**
Please read through as we should get the naming and wording right here.
This was initially discussed on gitter.

Resolves #22109

### CODE REVIEW

**Comment by jlprat** [comment]
I like the names, they are pretty explicit and clear what their intention is.

File: akka-actor/src/main/java/akka/annotation/ApiMayChange.java
Comment (jlprat) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: Should we also encourage to include some kind of Disclaimer as first line in the javadoc/scaladoc?

File: akka-actor/src/main/java/akka/annotation/ApiMayChange.java
Comment (patriknw) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: I think it should be allowed to even be changed in patch releases (2.5.x).

Also, mention something about that it's not guaranteed to be binary backwards compatibility.

Comment (jlprat) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: I know it's good to be able to change them even in patch releases, but would this then be too volatile for people to start trying them and giving useful feedback about those APIs?
If they can be completely changed from one patch to another, it might be not worth the effort to give it a try before it becomes stable

Comment (gosubpl) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: I think _technically_ minors (2.4, 2.5) are actually major releases for akka, so for me patches are equivalent of _minor_ releases, so I'd vote for patch releases

Comment (ktoso) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: > I think technically minors (2.4, 2.5) are actually major releases for akka

That's not true, see here: doc.akka.io/docs/akka/2.4/common/binary-compatibility-rules.html

File: akka-actor/src/main/java/akka/annotation/ApiMayChange.java
Comment (ktoso) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: Yeah will add

File: akka-actor/src/main/java/akka/annotation/ApiMayChange.java
Comment (gosubpl) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: From binary compatibility point yes, but I understand that we want to extend this with something in-between - I understand `ApiMayChange` as _this can change between minor releases contrary to general policy_ . My comment was to highlight that from my point of view the world for users of `akka` and (especially) `akka-http` is a fast moving one, and I (as a consumer of your product :) ) would like to be able to get updates to experimental apis faster than on minor release cycle

Comment (gosubpl) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: Ah, it seems that this looser binary compat policy was actually a case in the `epoch.major.minor` era :)

Comment (patriknw) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: We should not change if there is no good reason to change, and changing in a patch release should be done with care, such as deprecating and keeping the old if possible.

However, promising that we will be completely source compatible for all patch releases defeats the purpose of `ApiMayChange`. I would like to leave a door open for doing source incompatible changes also for patch releases, but it should be for exceptional cases.

Normally, I think, it's possible to iterate in patch releases in a source compatible way and then decide that now it's good and then finalize the api and remove the marker and the deprecated cruft. That final decision may also happen within a patch release cycle. The purpose is that we should be able to promote new modules to stable before releasing next 2.x.

Instead of promising things in the documentation related to version numbers it might be better to document this intention/process.

Comment (gosubpl) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: Yes, that totally makes sense. I was just trying to voice an opinion that `ApiMayChange` is exactly a way of allowing the change between minor releases *if absolutely necessary*

Comment (ktoso) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: I agree with @patriknw. What we're after is basically "-experimental" but within a stable module. We must be able to address feedback quickly and mark it stable in the next minor. E.g. thing comes in during 2.3.4, we iterate on it during 2.3.x, and promote it in 2.4.0 - no need to wait the full major cycle.

Comment (patriknw) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: It doesn't have to wait until 2.4.0 for promotion.

Thing comes in 2.3.4, we iterate in 2.3.5, 2.3.6, 2.3.7, and promote it in 2.3.8.

During 2.3.4 - 2.3.7 we do best effort in keeping it source compatible (exceptions may occur). In 2.3.8 it might not be source compatible, e.g. removing deprecated or doing final refactoring.

Comment (jlprat) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: Sounds good

Comment (ktoso) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: Ok, sounds good. No need to go as strict as I've outlined. Will amend the docs

Comment (ktoso) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: And add in the binary compat docs too

**Comment by ktoso** [comment]
Ready for another round: https://github.com/akka/akka/pull/22110/files#diff-188c9fc87aa0fab2b94d637d9d2cca0eR12

**Comment by gosubpl** [comment]
LGTM (If I may :) )

**Comment by jlprat** [comment]
Joining with LGTM if I may :) Just the comment in `@ApiMayChange` encouraging to add some line in scaladoc/javadoc stating the API may change

**Comment by ktoso** [comment]
Added, awaiting final LGTM

File: akka-actor/src/main/java/akka/annotation/ApiMayChange.java
Comment (patriknw) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: Perhaps also add:

It may also change when promoting the API to stable, for example such changes may include removal of deprecated methods that were introduced during the evolution and final refactoring that were deferred because they would have introduced to much breaking changes during the evolution phase. Promoting the API to stable may also happen in a patch release.

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-actor/src/main/java/akka/annotation/ApiMayChange.java
Comment (ktoso) [inline akka-actor/src/main/java/akka/annotation/ApiMayChange.java]: Well worded, will add and merge then

**Comment by ktoso** [comment]
Uff, 21 comments, less drama than I expected. Good!
Thanks everyone, these sound very good now.

**Comment by patriknw** [comment]
@ktoso remember to backport also

**Pull request merged**
